plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/OpenShiftProjectCache.java

```diff
@@ -0,0 +1,187 @@
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public synchronized void connectionChanged(IConnection connection, String property, Object oldValue, Object newValue) {
+		if (!(connection instanceof IOpenShiftConnection)) {
+			return;
+		}
+		if (ConnectionProperties.PROPERTY_PROJECTS.equals(property) && (oldValue instanceof List) && (newValue instanceof List)) {
+			List<IProject> newProjects = (List<IProject>) newValue;
+			List<IProject> oldProjects = (List<IProject>) oldValue;
+			IOpenShiftConnection conn = (IOpenShiftConnection)connection;
+			final String key = getCacheKey(conn);
+			
+			Set<IProjectAdapter> adapters = (Set<IProjectAdapter>) ObjectUtils.defaultIfNull(cache.get(key), new HashSet<>());
+			Map<IProject, IProjectAdapter> projectMap = adapters.stream().collect(Collectors.toMap(IProjectAdapter::getProject, Function.identity()));
+			
+			ListDiff diffs = Diffs.computeListDiff(oldProjects, newProjects);
+			Set<IProjectAdapter> added = new HashSet<>();
+			Set<IProjectAdapter> removed = new HashSet<>();
+			diffs.accept(new ListDiffVisitor() {
+				
+				@Override
+				public void handleRemove(int index, Object element) {
+					if(!(element instanceof IProject)) return;
+					IProject project = (IProject) element;
+					if(projectMap.containsKey(project)) {
+						IProjectAdapter adapter = projectMap.remove(project);
+						if(adapters.remove(adapter)) {
+							removed.add(adapter);
+						}
+					}
+				}
+				
+				@Override
+				public void handleAdd(int index, Object element) {
+					if(!(element instanceof IProject)) return;
+					IProjectAdapter adapter = newProjectAdapter(adapters, conn, (IProject)element);
+					if(adapter != null) {
+						added.add(adapter);
+					}
+				}
+			});
+			cache.put(key, adapters);
+			notifyRemove(removed);
+			notifyAdd(added);
+		}
+	}
+	
+	
+	
+	@Override
+	public void connectionAdded(IConnection connection) {
+	}
+
+	@Override
+	public void connectionRemoved(IConnection connection) {
+		if (!(connection instanceof IOpenShiftConnection)) {
+			return;
+		}
+		IOpenShiftConnection conn = (IOpenShiftConnection) connection;
+		flushFor(conn);
+	}
+
+	private IProjectAdapter newProjectAdapter(Collection<IProjectAdapter> adapters, IOpenShiftConnection conn, IProject project) {
+		OpenShiftProjectUIModel model = new OpenShiftProjectUIModel(conn, project);
+		if(adapters.add(model)) {;
+			return model;
+		}
+		return null;
+	}
+	
+	
+	private void notifyAdd(Collection<IProjectAdapter> adapters) {
+		synchronized (listeners) {
+			try {
+				for (IProjectAdapter a : adapters) {
+					listeners.forEach(l->l.handleAddToCache(this, a));
+				}
+			}catch(Exception e) {
+				Trace.error("Exception while trying to notify cache listener", e);
+			}
+		}
+	}
+
+	private void notifyRemove(Collection<IProjectAdapter> adapters) {
+		synchronized (listeners) {
+			try {
+				for (IProjectAdapter a : adapters) {
+					listeners.forEach(l->l.handleRemoveFromCache(this, a));
```

Comment:
@jcantrill Cache listeners are notified of removed IProjectAdapter and they are already removed from 'cache' map. That is fine. However, DeploymentResourceMapper is still a listener to ConnectionsRegistry, so that it cannot be handled by garbage collector together with all maps that it built. I suggest
{code}
public void IProjectAdapter.dispose() {
    mapper.dispose();
}

public void DeploymentResourceMapper.dispose() {
    ConnectionsRegistrySingleton.getInstance().removeListener(connectionListener);
    cache.dispose();
    //clean maps for faster memory cleaning.
}

public void ObservableResourceCache.dispose() {
    //clean maps and collections for faster memory cleaning.
}

{code}